chia/full_node/full_node.py

```diff
@@ -1026,19 +1026,54 @@ async def respond_block(
             return None
 
         pre_validation_result: Optional[PreValidationResult] = None
-        if block.is_transaction_block() and block.transactions_generator is None:
+        if (
+            block.is_transaction_block()
+            and block.transactions_info is not None
+            and block.transactions_info.generator_root != bytes([0] * 32)
+            and block.transactions_generator is None
+        ):
             # This is the case where we already had the unfinished block, and asked for this block without
             # the transactions (since we already had them). Therefore, here we add the transactions.
             unfinished_rh: bytes32 = block.reward_chain_block.get_unfinished().get_hash()
             unf_block: Optional[UnfinishedBlock] = self.full_node_store.get_unfinished_block(unfinished_rh)
-            if unf_block is not None and unf_block.transactions_generator is not None:
+            if (
+                unf_block is not None
+                and unf_block.transactions_generator is not None
+                and unf_block.foliage_transaction_block == block.foliage_transaction_block
+            ):
                 pre_validation_result = self.full_node_store.get_unfinished_block_result(unfinished_rh)
                 assert pre_validation_result is not None
                 block = dataclasses.replace(
                     block,
                     transactions_generator=unf_block.transactions_generator,
                     transactions_generator_ref_list=unf_block.transactions_generator_ref_list,
                 )
+            else:
+                # We still do not have the correct information for this block, perhaps there is a duplicate block
+                # with the same unfinished block hash in the cache, so we need to fetch the correct one
+                if peer is None:
+                    return None
+
+                block_response: Optional[Any] = await peer.request_block(
+                    full_node_protocol.RequestBlock(block.height, True)
+                )
+                if block_response is None or not isinstance(block_response, full_node_protocol.RespondBlock):
+                    self.log.warning(
+                        f"Was not able to fetch the correct block for height {block.height} {block_response}"
+                    )
+                    return None
+                new_block: FullBlock = block_response.block
+                if new_block.foliage_transaction_block != block.foliage_transaction_block:
+                    self.log.warning(f"Received the wrong block for height {block.height} {new_block.header_hash}")
+                    return None
+                assert new_block.transactions_generator is not None
```

Comment:
what if the peer sends back a block without the transactions? it seems like an assert may be too strong of an assumption here. Or do we check this when we receive the response?
I'm trying to think of ways for a malicious peer to force us to keep recursing, but as long as this holds, I don't think it's possible

Reply:
it will throw AssertionError and disconnect the peer

Reply:
The `if new_block.foliage_transaction_block != block.foliage_transaction_block` should return if its an honest peer. Attacker peers would get disconnected in the assertion

Reply:
I wouldn't expect to test conditions that can be false in `assert`, but only check internal consistency that way. It's not a big deal, but I think it would be a bit nicer to have a regular check for this condition (maybe even with a comment explaining why it's there)